File: src/main/java/core/pipeline/impl/ConsumerBlock.java
package core.pipeline.impl;

import core.data.Data;
import core.pipeline.PipelineBlock;

import java.util.function.Consumer;


/**
 *  A {@link PipelineBlock}, that is only executing {@link Consumer<Data>} provided.
 *  Subscription to this block is not allowed and will lead to an error.
 */
public class ConsumerBlock extends PipelineBlock<Data, Data> {

    private Consumer<Data> consumer;

    public ConsumerBlock(Consumer<Data> consumer) {
        this.consumer = consumer;
    }

    @Override
    public void InputDataReady(PipelineBlock<?, Data> sender, Data data) {
        consumer.accept(data);
    }

    @Override
    public void addListener(PipelineBlock<Data, ?> listener) {
        throw new RuntimeException("Subscription to the ConsumerBlock block is not allowed");
    }
}

File: src/main/java/core/pipeline/impl/ConsumerBlock.kt
package core.pipeline.impl

import core.data.Data
import core.pipeline.PipelineBlock

import java.util.function.Consumer


/**
 * A [PipelineBlock], that is only executing [<] provided.
 * Subscription to this block is not allowed and will lead to an error.
 */
class ConsumerBlock(private val consumer: Consumer<Data>) : PipelineBlock<Data, Data>() {

    override fun flush() {}

    override fun inputDataReady(sender: PipelineBlock<*, Data>?, input: Data) {
        consumer.accept(input)
    }

    override fun addListener(listener: PipelineBlock<Data, *>) {
        throw RuntimeException("Subscription to the ConsumerBlock block is not allowed")
    }
}
